### PR TITLE
Support for Tracing Headers

### PR DESCRIPTION
**Pull request by spolu**
r? @zenazn

### CODE REVIEW

File: lib/chalk-log/layout.rb [hunk 1]
```diff
@@ -164,20 +164,35 @@ def json(value)
...
+        :action_id => LSpace[:action_id],
+        :span_id => LSpace[:span_id] || 0,
+        :parent_id => LSpace[:parent_id] || 0
+      }
+    end
+    return id
+  end
+
+  def tracing_id_prefix(tracing_id)
+    tracing_id[:action_id] + " " +
```
Comment (zenazn) [inline lib/chalk-log/layout.rb]: Ruby also has `sprintf`

File: lib/chalk-log/layout.rb [hunk 2]
```diff
@@ -164,20 +164,35 @@ def json(value)
...
-
   def contextual_info
     LSpace[:'chalk.log.contextual_info']
   end
 
-  def tag(message, time, pid, action_id)
+  def tracing_id
+    id = nil
+    if !LSpace[:action_id].nil?
+      id = {
```
Comment (zenazn) [inline lib/chalk-log/layout.rb]: Can we make this a `Struct`?

File: lib/chalk-log/layout.rb [hunk 3]
```diff
@@ -164,20 +164,35 @@ def json(value)
...
+  end
+
+
+  def tag(message, time, pid, tracing_id)
     return message unless configatron.chalk.log.tagging
 
     metadata = []
     metadata << pid if configatron.chalk.log.pid
-    metadata << action_id if action_id
+    metadata << tracing_id_prefix(tracing_id) if tracing_id
```
Comment (zenazn) [inline lib/chalk-log/layout.rb]: can we make this a `to_s` on the tracing id class

File: lib/chalk-log/layout.rb [hunk 4]
```diff
@@ -164,20 +164,35 @@ def json(value)
...
     LSpace[:'chalk.log.contextual_info']
   end
 
-  def tag(message, time, pid, action_id)
+  def tracing_id
+    id = nil
+    if !LSpace[:action_id].nil?
+      id = {
+        :action_id => LSpace[:action_id],
+        :span_id => LSpace[:span_id] || 0,
```
Comment (zenazn) [inline lib/chalk-log/layout.rb]: This isn't right—we need to generate a new random int

Comment (zenazn) [inline lib/chalk-log/layout.rb]: (we should duplicate the logic in package `span` exactly)

**Comment by zenazn** [comment]
Actually I thiiink we're handling this at the wrong level of abstraction: the "action ID / span ID" object should be handed to chalk-log wholesale, and should understand how to render itself (i.e., no rendering logic in chalk-log). That way others can plug their format of choice into chalk-log, for instance

**Comment by spolu** [comment]
So if I understand correctly use `.to_s` on whatever is present in `LSpace[:span_id]` ?

**Comment by zenazn** [comment]
Yeah, I think so (and falling back to `:action_id` or whatever if it's not set)

**Comment by spolu** [comment]
Sounds reasonable indeed

**Comment by spolu** [comment]
Updated

File: lib/chalk-log/layout.rb [hunk 5]
```diff
@@ -164,20 +163,23 @@ def json(value)
...
-  def action_id
-    LSpace[:action_id]
-  end
-
   def contextual_info
     LSpace[:'chalk.log.contextual_info']
   end
 
-  def tag(message, time, pid, action_id)
+  def span
```
Comment (zenazn) [inline lib/chalk-log/layout.rb]: `LSpace[:action_id] || LSpace[:span]`

**Comment by zenazn** [comment]
LGTM :love_letter:

**Pull request merged**